main/Repositories/Database/ChatRepository.cs

```diff
@@ -0,0 +1,112 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using NuGet.Versioning;
+using ValuedInBE.Contexts;
+using ValuedInBE.Models;
+using ValuedInBE.Models.Entities.Messaging;
+using ValuedInBE.System.Extensions;
+using ValuedInBE.System.Middleware;
+
+namespace ValuedInBE.Repositories.Database
+{
+    public class ChatRepository : IChatRepository
+    {
+        private readonly ValuedInContext _context;
+        private readonly IHttpContextAccessor _contextAccessor;
+
+
+        public ChatRepository(ValuedInContext context, IHttpContextAccessor contextAccessor)
+        {
+            _context = context;
+            _contextAccessor = contextAccessor;
+        }
+
+        public async Task AddChatParticipantsAsync(List<ChatParticipant> participants)
+        {
+            _context.ChatParticipants.AddRange(participants);
+            CheckEntityAuditing();
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task SaveNewChatAsync(Chat chat)
+        {
+            _context.Chats.Add(chat);
+            CheckEntityAuditing();
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task CreateNewChatMessageAsync(ChatMessage chatMessage)
+        {
+            _context.ChatMessages.Add(chatMessage);
+            CheckEntityAuditing();
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task<Chat> GetChatFromParticipantsAsync(List<string> allParticipants)
+        {
+            var a = from c in _context.Chats
+                                .Where(c => c.Participants.Any() &&
+                                         c.Participants.All(p => allParticipants.Contains(p.UserId))
+                                )
+                    select c;
+
+            return await a.FirstOrDefaultAsync();
+        }
+
+        public async Task<IEnumerable<Chat>> GetChatsWithLastMessageAndParticipantsAsync(string userId, int pageSize, DateTime? createdSince)
+        {
+            UserContext userContext = _contextAccessor.HttpContext.GetUserContext();
+            var a = from c in _context.Chats
+                                .Include(c=>c.Messages.OrderByDescending(m => m.CreatedOn).Take(1))
+                                .Include(c=>c.Participants)
+                                .Take(pageSize)
+                                .Where(c=>
+                                    c.Participants.Select(p=>p.UserId).Contains(userContext.UserID)
+                                    && (createdSince.HasValue && c.CreatedOn > createdSince)
```

Comment:
createdSince as a parameter can be null, will everything be okay in "c.CreatedOn > createdSince" check if it is null?

Reply:
createdSince.HasValue verifies that it has value